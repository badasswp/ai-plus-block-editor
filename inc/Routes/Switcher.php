<?php
/**
 * Switcher Route.
 *
 * This route is responsible for updating the
 * AI Provider.
 *
 * @package AiPlusBlockEditor
 */

namespace AiPlusBlockEditor\Routes;

use AiPlusBlockEditor\Core\AI;
use AiPlusBlockEditor\Admin\Options;
use AiPlusBlockEditor\Abstracts\Route;
use AiPlusBlockEditor\Interfaces\Router;

/**
 * Switcher class.
 */
class Switcher extends Route implements Router {
	/**
	 * Get, Post, Put, Patch, Delete.
	 *
	 * @since 1.5.0
	 *
	 * @var string
	 */
	public string $method = 'POST';

	/**
	 * WP REST Endpoint e.g. /wp-json/ai-plus-block-editor/v1/switcher.
	 *
	 * @since 1.5.0
	 *
	 * @var string
	 */
	public string $endpoint = '/switcher';

	/**
	 * WP_REST_Request object.
	 *
	 * @since 1.5.0
	 *
	 * @var \WP_REST_Request
	 */
	public \WP_REST_Request $request;

	/**
	 * Response Callback.
	 *
	 * @since 1.5.0
	 *
	 * @return \WP_REST_Response|\WP_Error
	 */
	public function response() {
		$args = $this->request->get_json_params();

		if ( empty( $args['provider'] ?? '' ) ) {
			return $this->get_400_response(
				sprintf(
					'API Request does not contain a valid AI provider. Provider: %s',
					$args['provider'] ?? ''
				)
			);
		}

		update_option(
			Options::get_page_option(),
			wp_parse_args(
				[
					'ai_provider' => sanitize_text_field( $args['provider'] ),
				],
				get_option( Options::get_page_option(), [] )
			)
		);

		return rest_ensure_response(
			[
				'message'  => sprintf(
					'AI Provider switched successfully to %s',
					$args['provider']
				),
				'provider' => get_option( Options::get_page_option(), [] )['provider'] ?? '',
			]
		);
	}
}
