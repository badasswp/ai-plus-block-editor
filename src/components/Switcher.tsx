import { useState } from '@wordpress/element';
import { useDispatch } from '@wordpress/data';
import { __, sprintf } from '@wordpress/i18n';
import { SelectControl } from '@wordpress/components';
import { store as noticesStore } from '@wordpress/notices';
import apiFetch from '@wordpress/api-fetch';

/**
 * Switcher.
 *
 * This Component handles the switching of AI providers,
 * for e.g. Open AI -> Gemini and so on.
 *
 * @since 1.5.0
 *
 * @return {JSX.Element} Switcher.
 */
const Switcher = (): JSX.Element => {
	const { createNotice } = useDispatch( noticesStore );
	const [ provider, setProvider ] = useState< string >( apbe.provider || '' );

	const handleChange = async ( value: string ) => {
		setProvider( value );

		try {
			await apiFetch( {
				path: '/ai-plus-block-editor/v1/switcher',
				method: 'POST',
				data: {
					provider: value,
				},
			} );

			createNotice(
				'success',
				__(
					'Success! AI Provider switched successfully.',
					'ai-plus-block-editor'
				),
				{
					isDismissible: true,
					id: 'apbe-success',
					type: 'snackbar',
				}
			);
		} catch ( e ) {
			createNotice(
				'error',
				__(
					'Error! Failed to switch AI Provider.',
					'ai-plus-block-editor'
				),
				{
					isDismissible: true,
					id: 'apbe-error',
					type: 'snackbar',
				}
			);
			// eslint-disable-next-line no-console
			console.error( e.message );
		}
	};

	return (
		<>
			<p>
				<strong>
					{ __( 'Switch AI Provider', 'ai-plus-block-editor' ) }
				</strong>
			</p>
			<SelectControl
				label={ null }
				value={ provider }
				options={ [
					{ label: 'ChatGPT', value: 'OpenAI' },
					{ label: 'Gemini', value: 'Gemini' },
					{ label: 'DeepSeek', value: 'DeepSeek' },
				] }
				onChange={ handleChange }
				id="switcher"
				__next40pxDefaultSize
				__nextHasNoMarginBottom
			/>
		</>
	);
};

export default Switcher;
