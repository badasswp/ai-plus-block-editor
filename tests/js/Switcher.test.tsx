import '@testing-library/jest-dom';
import { fireEvent, render } from '@testing-library/react';
import { useDispatch } from '@wordpress/data';
import Switcher from '../../src/components/Switcher';

jest.mock( '@wordpress/i18n', () => ( {
	__: jest.fn( ( arg ) => arg ),
	sprintf: jest.fn(),
} ) );

jest.mock( '@wordpress/data', () => ( {
	useSelect: jest.fn(),
	useDispatch: jest.fn(),
} ) );

jest.mock( '@wordpress/notices', () => ( {
	store: 'core/notices',
} ) );

jest.mock( '@wordpress/components', () => ( {
	SelectControl: jest.fn( ( { label, value, options, onChange, id } ) => {
		return (
			<>
				<p>{ label }</p>
				<select
					onChange={ ( e ) => {
						const selectedText =
							e.target.options[ e.target.selectedIndex ].text;
						onChange( selectedText );
					} }
					value={ value }
					data-testid={ id }
				>
					{ options.map(
						(
							item: { label: string; value: string },
							index: number
						) => {
							return (
								<option key={ index } value={ item.value }>
									{ item.label }
								</option>
							);
						}
					) }
				</select>
			</>
		);
	} ),
} ) );

describe( 'Switcher', () => {
	beforeEach( () => {
		( useDispatch as jest.Mock ).mockReturnValue( {
			createNotice: jest.fn(),
		} );

		( window as any ).apbe = {
			provider: 'OpenAI',
		};
	} );

	afterEach( () => {
		jest.clearAllMocks();
	} );

	it( 'renders Switcher component', () => {
		const { container, getByText, getByRole } = render( <Switcher /> );

		expect( container ).toMatchSnapshot();

		expect( getByText( 'Switch AI Provider' ) ).toBeVisible();
		expect( getByRole( 'combobox' ) ).toBeVisible();
		expect( getByText( 'ChatGPT' ) ).toBeVisible();
		expect( getByText( 'Gemini' ) ).toBeVisible();
		expect( getByText( 'DeepSeek' ) ).toBeVisible();
	} );

	it( 'makes an API call request on selection change', () => {
		const mockCreateNotice = jest.fn();

		( useDispatch as jest.Mock ).mockReturnValue( {
			createNotice: mockCreateNotice,
		} );

		const { getByTestId } = render( <Switcher /> );

		const select = getByTestId( 'switcher' );
		fireEvent.change( select, { target: { value: 'Gemini' } } );

		expect( useDispatch ).toHaveBeenCalledTimes( 2 );
	} );
} );
