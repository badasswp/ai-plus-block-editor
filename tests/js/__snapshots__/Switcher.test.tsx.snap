// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Switcher renders Switcher component 1`] = `
<div>
  <p>
    <strong>
      Switch AI Provider
    </strong>
  </p>
  <p />
  <select
    data-testid="switcher"
  >
    <option
      value="OpenAI"
    >
      ChatGPT
    </option>
    <option
      value="Gemini"
    >
      Gemini
    </option>
    <option
      value="DeepSeek"
    >
      DeepSeek
    </option>
    <option
      value="Grok"
    >
      Grok
    </option>
  </select>
</div>
`;
